src/VisualStudio/Core/Impl/SolutionExplorer/DiagnosticItem/BaseDiagnosticAndGeneratorItemSource.cs

```diff
@@ -7,187 +7,208 @@
-                AnalyzerReference.GetGenerators(language)
-                .Select(g => new SourceGeneratorItem(projectId, g, AnalyzerReference)));
+                    var effectiveSeverity = selectedDiagnostic.GetEffectiveSeverity(
+                        project.CompilationOptions!,
+                        analyzerConfigOptions?.ConfigOptions,
+                        analyzerConfigOptions?.TreeOptions);
+                    return (BaseItem)new DiagnosticItem(project.Id, analyzerReference, selectedDiagnostic, effectiveSeverity, CommandHandler);
+                });
+        }
 
-            return collection;
+        async Task<ImmutableArray<BaseItem>> GenerateSourceGeneratorItemsAsync(
+            Project project,
+            AnalyzerReference analyzerReference)
+        {
+            var identifies = await GetIdentitiesAsync().ConfigureAwait(false);
+            return identifies.SelectAsArray(
+                identity => (BaseItem)new SourceGeneratorItem(project.Id, identity, analyzerReference.FullPath));
         }
 
-        private void OnWorkspaceChangedLookForOptionsChanges(object sender, WorkspaceChangeEventArgs e)
+        async Task<ImmutableArray<SourceGeneratorIdentity>> GetIdentitiesAsync()
         {
-            if (e.Kind is WorkspaceChangeKind.SolutionCleared or
-                WorkspaceChangeKind.SolutionReloaded or
-                WorkspaceChangeKind.SolutionRemoved)
-            {
-                Workspace.WorkspaceChanged -= OnWorkspaceChangedLookForOptionsChanges;
-            }
-            else if (e.ProjectId == ProjectId)
-            {
-                if (e.Kind == WorkspaceChangeKind.ProjectRemoved)
-                {
-                    Workspace.WorkspaceChanged -= OnWorkspaceChangedLookForOptionsChanges;
-                }
-                else if (e.Kind == WorkspaceChangeKind.ProjectChanged)
-                {
-                    OnProjectConfigurationChanged();
-                }
-                else if (e.DocumentId != null)
-                {
-                    switch (e.Kind)
-                    {
-                        case WorkspaceChangeKind.AnalyzerConfigDocumentAdded:
-                        case WorkspaceChangeKind.AnalyzerConfigDocumentChanged:
-                        case WorkspaceChangeKind.AnalyzerConfigDocumentReloaded:
-                        case WorkspaceChangeKind.AnalyzerConfigDocumentRemoved:
-                            OnProjectConfigurationChanged();
-                            break;
-                    }
-                }
-            }
+            // Can only remote AnalyzerFileReferences over to the oop side.
+            if (analyzerReference is not AnalyzerFileReference analyzerFileReference)
+                return SourceGeneratorIdentity.GetIdentities(analyzerReference, project.Language);
 
-            return;
+            var client = await RemoteHostClient.TryGetClientAsync(this.Workspace, cancellationToken).ConfigureAwait(false);
```

Comment:
new code that now calls into OOP to compute things for the normal VS case of an AnalyzerFileReference.